### ISSUE TITLE
Migrations ignore default schema

### ISSUE THREAD

**Issue by ghost** [dotnet/efcore#9864]
For testing purposes I try to apply a set of migrations to a different schema for each test run (or in production for a new installation into a customer-supplied schema, for that matter). For that purposes I have a `_schema` field in my database context and use it to set the default schema on the ModelBuilder in `OnModelCreating()`. In my migrations all tables are in the standard schema, i.e. don't have a schema-field set. When applying the migrations via `ctx.Database.Migrate()` (ctx being the context instance with the default schema set), tables are still created in the dbo.-schema on MSSQL. Subsequent access to the tables obviously fails, as the context is looking for the tables in the configured default schema.

When I create a new migration from a context with the default schema set, each table receives the correct schema field, so `Add-Migration` seems to obey the schema. `ctx.Database.Migrate()` should do so as well for consistency reasons (at least for tables, that have no explicit schema set). 

### Further technical details
EF Core version: 2.0.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017

**Comment by bricelam** [comment]
Migrations without a schema specified will generate SQL without a schema specified. When executing, it will use the database user's default schema.

You can change the user's default schema:
``` sql
ALTER USER MyUser WITH DEFAULT_SCHEMA = MySchema
```

You could also update you migrations to use a custom base type to flow the default schema in:

``` csharp
class MyMigrationBase : Migration
{
    public static string DefaultSchema { get; set; }
}
```

``` csharp
class MyMigration1 : MyMigrationBase
{
    protected override void Up(MigraitonBuilder builder)
    {
        builder.DropColumn(
            schema: DefaultSchema,
            table: "MyTable",
            column: "MyColumn");
    }
}
```
...then set it before migrating.
``` csharp
MyMigrationBase.DefaultSchema = ctx.Model.Relational().DefaultSchema;
ctx.Database.Migrate();

**Comment by markusbrueckner** [comment]
While this seems to work (although it is quite the effort to change all points in all migrations to include the schema), I'm now stuck with the issue of the IModel being cached. The migrations create the correct schema, but using the DbContext re-uses the wrong model after the first test case. Therefore all Entities point to the wrong schema, making the test case fail again. I'll probably have to go with some kind of multi-tenancy approach there, I guess?

**Comment by ajcvickers** [comment]
@markusbrueckner You could look into providing an override of the `ModelCacheKeyFactory` and registering it with `ReplaceService` in DbContextOptionsBuilder. Beyond this, we are leaving this issue open to consider some other things we might do with schemas that could be better for this kind of scenario. Will discuss with @divega when he is back.

**Comment by markusbrueckner** [comment]
Thanks for the advice. I implemented an IModelCacheFactory that uses a Schema-field of my database context into account. So technically things work now, but performance is abysmal. My problem is: I'm creating and destroying the tables for test cases. Since the `DefaultSchema` field in the approach above uses a static field, I have to protect this using a lock. This effectively prevents the tests from running in parallel, which is a problem if your test numbers are in the three digits. 

I'm coming to think that my approach is flawed from the start. I tried with `EnsureCreated/EnsureDeleted` on different LocalDBs first, but that was unacceptable (taking approx. 15-10 sec. to  create a DB). The schema-approach is much faster for individual schemas, but doesn't scale due to the parallelism being lost. I cannot use the InMemory-DB or SQLite, even though they show decent performance. They simply do not support features needed for my model. So in the end I'm looking for a fast and scalable way to isolate automated tests. Not sure whether schemas can be a way for that.
